crates/rune-modules/src/net.rs

```diff
@@ -0,0 +1,185 @@
+    module.ty::<IpAddr>()?;
+
+    module.function_meta(SocketAddr::new__meta)?;
+    module.function_meta(SocketAddr::ip__meta)?;
+    module.function_meta(SocketAddr::set_ip__meta)?;
+    module.function_meta(SocketAddr::port__meta)?;
+    module.function_meta(SocketAddr::set_port__meta)?;
+    module.function_meta(SocketAddr::is_ipv4__meta)?;
+    module.function_meta(SocketAddr::is_ipv6__meta)?;
+
+    module.function_meta(IpAddr::is_unspecified__meta)?;
+    module.function_meta(IpAddr::is_loopback__meta)?;
+    module.function_meta(IpAddr::is_multicast__meta)?;
+    module.function_meta(IpAddr::is_ipv4__meta)?;
+    module.function_meta(IpAddr::is_ipv6__meta)?;
+    module.function_meta(IpAddr::to_canonical__meta)?;
+
+    Ok(module)
+}
+
+/// An internet socket address, either IPv4 or IPv6.
+#[derive(Debug, Any)]
+#[rune(item = ::net)]
+pub struct SocketAddr {
+    inner: net::SocketAddr,
+}
+
+impl SocketAddr {
+    /// Creates a new socket address from an IP address and a port number.
+    #[rune::function(keep, path = Self::new)]
+    pub const fn new(ip: IpAddr, port: u16) -> Self {
+        Self {
+            inner: net::SocketAddr::new(ip.inner, port),
+        }
+    }
+
+    /// Returns the IP address associated with this socket address.
+    #[rune::function(instance, keep)]
+    pub const fn ip(&self) -> IpAddr {
+        IpAddr {
+            inner: self.inner.ip(),
+        }
+    }
+
+    /// Changes the IP address associated with this socket address.
+    #[rune::function(instance, keep)]
+    pub fn set_ip(&mut self, new_ip: IpAddr) {
+        self.inner.set_ip(new_ip.inner);
+    }
+
+    /// Returns the port number associated with this socket address.
+    #[rune::function(instance, keep)]
+    pub const fn port(&self) -> u16 {
+        self.inner.port()
+    }
+
+    /// Changes the port number associated with this socket address.
+    #[rune::function(instance, keep)]
+    pub fn set_port(&mut self, new_port: u16) {
+        self.inner.set_port(new_port);
+    }
+
+    /// Returns [`true`] if the IP address in this `SocketAddr` is an
+    /// `IPv4` address, and [`false`] otherwise.
+    #[rune::function(instance, keep)]
+    pub const fn is_ipv4(&self) -> bool {
+        self.inner.is_ipv4()
+    }
+
+    /// Returns [`true`] if the IP address in this `SocketAddr` is an
+    /// `IPv6` address, and [`false`] otherwise.
+    #[rune::function(instance, keep)]
+    pub const fn is_ipv6(&self) -> bool {
+        self.inner.is_ipv6()
+    }
+}
+
+impl SocketAddr {
+    /// Converts [`SocketAddr`] into a [`std::net::SocketAddr`].
+    pub const fn into_std(self) -> net::SocketAddr {
+        self.inner
+    }
+
+    /// Creates a [`SocketAddr`] from a [`std::net::SocketAddr`].
+    pub const fn from_std(addr: net::SocketAddr) -> Self {
+        Self { inner: addr }
+    }
+}
+
+/// An IP address, either IPv4 or IPv6.
+#[derive(Debug, Any)]
+#[rune(item = ::std::net)]
+pub struct IpAddr {
+    inner: net::IpAddr,
+}
```

Comment:
We need a way to create an `IpAddr`. What is a preference: enum or constructors?

The std is `IpAddr::V4(...)` or `V6(...)`